File: ios/Classes/BucketeerFlutterClientSdkPlugin.swift
import Flutter
import UIKit
import Bucketeer

public class BucketeerFlutterClientSdkPlugin: NSObject, FlutterPlugin {

    private static let METHOD_CHANNEL_NAME = "io.bucketeer.sdk.plugin.flutter"
    private static let EVALUATION_UPDATE_EVENT_CHANNEL_NAME = "\(METHOD_CHANNEL_NAME)::evaluation.update.listener"

    private let logger = BucketeerPluginLogger()
    private let proxyEvaluationListener = BucketeerPluginEvaluationUpdateListener()
    private var proxyEvaluationListenToken: String?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: METHOD_CHANNEL_NAME, binaryMessenger: registrar.messenger())
        let instance = BucketeerFlutterClientSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        let eventChannel = FlutterEventChannel(name: EVALUATION_UPDATE_EVENT_CHANNEL_NAME,
                                               binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance.proxyEvaluationListener)
    }

    private func initialize(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let apiKey = arguments?["apiKey"] as? String else {
            fail(result: result, message: "apiKey is required")
            return
        }
        guard let apiEndpoint = arguments?["apiEndpoint"] as? String else {
            fail(result: result, message: "apiEndpoint is required")
            return
        }
        guard let userId = arguments?["userId"] as? String else {
            fail(result: result, message: "userId is required")
            return
        }
        guard let appVersion = arguments?["appVersion"] as? String else {
            fail(result: result, message: "appVersion is required")
            return
        }
        
        let featureTag = (arguments?["featureTag"] as? String) ?? ""
        
        do {
            var builder = BKTConfig.Builder()
                .with(apiKey: apiKey)
                .with(apiEndpoint: apiEndpoint)
                .with(featureTag: featureTag)
                .with(appVersion: appVersion)

            if let eventsFlushInterval = arguments?["eventsFlushInterval"] as? Int64 {
                builder = builder.with(eventsFlushInterval: eventsFlushInterval)
            }

            if let eventsMaxQueueSize = arguments?["eventsMaxQueueSize"] as? Int {
                builder = builder.with(eventsMaxQueueSize: eventsMaxQueueSize)
            }

            if let pollingInterval = arguments?["pollingInterval"] as? Int64 {
                builder = builder.with(pollingInterval: pollingInterval)
            }

            if let backgroundPollingInterval = arguments?["backgroundPollingInterval"] as? Int64 {
                builder = builder.with(backgroundPollingInterval: backgroundPollingInterval)
            }

            if let debugging = arguments?["debugging"] as? Bool, debugging {
                builder = builder.with(logger: logger)
            }

            let bkConfig = try builder.build()
            let userAttributes = arguments?["userAttributes"] as? [String: String] ?? [:]
            let user = try BKTUser.Builder().with(id: userId).with( attributes: userAttributes).build()
            
            let completion : ((BKTError?) -> Void) = { [self] err in
                if let er = err {
                    if case .timeout(_, _, _) = err {
                        logger.warn(message: "Fetch evaluations failed during the initialize process. It will try to fetch again in the next polling.")
                    } else {
                        logger.error(message: "BKTClient.initialize failed with error: \(er)", er)
                        fail(result: result, message: er.localizedDescription)
                    }
                } else {
                    success(result: result)
                }
            }
                        
            if let timeoutMillis = arguments?["timeoutMillis"] as? Int64 {
                try BKTClient.initialize(config: bkConfig, user: user, timeoutMillis: timeoutMillis, completion: completion)
            } else {
                try BKTClient.initialize(config: bkConfig, user: user, completion: completion)
            }

        } catch {
            logger.error(message: "BKTClient.initialize failed with error: \(error)", error)
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func stringVariation(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let featureId = arguments?["featureId"] as? String else {
            fail(result: result, message: "featureId is required")
            return
        }
        guard let defaultValue = arguments?["defaultValue"] as? String else {
            fail(result: result, message: "defaultValue is required")
            return
        }
        do {
            let response = try BKTClient.shared.stringVariation(featureId: featureId, defaultValue: defaultValue)
            success(result: result, response: response)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func intVariation(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let featureId = arguments?["featureId"] as? String else {
            fail(result: result, message: "featureId is required")
            return
        }
        guard let defaultValue = arguments?["defaultValue"] as? Int else {
            fail(result: result, message: "defaultValue is required")
            return
        }
        do {
            let response = try BKTClient.shared.intVariation(featureId: featureId, defaultValue: defaultValue)
            success(result: result, response: response)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func doubleVariation(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let featureId = arguments?["featureId"] as? String else {
            fail(result: result, message: "featureId is required")
            return
        }
        guard let defaultValue = arguments?["defaultValue"] as? Double else {
            fail(result: result, message: "defaultValue is required")
            return
        }
        do {
            let response = try BKTClient.shared.doubleVariation(featureId: featureId, defaultValue: defaultValue)
            success(result: result, response: response)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func boolVariation(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let featureId = arguments?["featureId"] as? String else {
            fail(result: result, message: "featureId is required")
            return
        }
        guard let defaultValue = arguments?["defaultValue"] as? Bool else {
            fail(result: result, message: "defaultValue is required")
            return
        }
        do {
            let response = try BKTClient.shared.boolVariation(featureId: featureId, defaultValue: defaultValue)
            success(result: result, response: response)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func track(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let goalId = arguments?["goalId"] as? String else {
            fail(result: result, message: "goalId is required")
            return
        }
        do {
            if let value = arguments?["value"] as? Double {
                try BKTClient.shared.track(goalId: goalId, value: value)
            } else {
                try BKTClient.shared.track(goalId: goalId)
            }
            success(result: result)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func jsonVariation(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let featureId = arguments?["featureId"] as? String else {
            fail(result: result, message: "featureId is required")
            return
        }
        guard let defaultValue = arguments?["defaultValue"] as? [String: AnyHashable] else {
            fail(result: result, message: "defaultValue is required")
            return
        }
        do {
            let response = try BKTClient.shared.jsonVariation(featureId: featureId, defaultValue: defaultValue)
            success(result: result, response: response)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func currentUser(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        var client: BKTClient?
        do {
            client = try BKTClient.shared
        } catch {
            fail(result: result, message: error.localizedDescription)
            return
        }

        guard let user = client?.currentUser() else {
            fail(result: result, message: "Failed to fetch the user.")
            return
        }
        success(result: result, response: ["id": user.id, "data": user.attr])
    }

    private func updateUserAttributes(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let userAttributes = arguments as? [String: String] else {
            fail(result: result, message: "userAttributes is required")
            return
        }
        do {
            try BKTClient.shared.updateUserAttributes(attributes: userAttributes)
            success(result: result)
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func fetchEvaluations(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        let timeoutMillis = arguments?["timeoutMillis"] as? Int64
        do {
            try BKTClient.shared.fetchEvaluations(timeoutMillis: timeoutMillis) { [weak self] err in
                if let err {
                    self?.fail(result: result, message: err.localizedDescription)
                } else {
                    self?.success(result: result)
                }
            }
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func flush(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        do {
            try BKTClient.shared.flush {[weak self] error in
                if let bktError = error {
                    let errorMessage = bktError.localizedDescription
                    self?.fail(result: result, message: errorMessage)
                } else {
                    self?.success(result: result)
                }
            }
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }

    private func evaluationDetails(_ arguments: [String: Any]?, _ result: @escaping FlutterResult) {
        guard let featureId = arguments?["featureId"] as? String else {
            fail(result: result, message: "featureId is required")
            return
        }
        var client: BKTClient?
        do {
            client = try BKTClient.shared
        } catch {
            fail(result: result, message: error.localizedDescription)
            return
        }

        guard let response = client?.evaluationDetails(featureId: featureId) else {
            fail(result: result, message: "Feature flag not found.")
            return
        }
        success(
            result: result,
            response: [
                "id": response.id,
                "featureId": response.featureId,
                "featureVersion": response.featureVersion,
                "userId": response.userId,
                "variationId": response.variationId,
                "variationName": response.variationName,
                "variationValue": response.variationValue,
                "reason": response.reason.rawValue
            ])
    }

    private func addProxyEvaluationUpdateListener(_ result: @escaping FlutterResult) {
        do {
            if let listenToken = proxyEvaluationListenToken {
                success(result: result, response: listenToken)
            } else {
                let newListenToken = try BKTClient.shared.addEvaluationUpdateListener(listener: proxyEvaluationListener)
                proxyEvaluationListenToken = newListenToken
                success(result: result, response: newListenToken)
            }
        } catch {
            fail(result: result, message: error.localizedDescription)
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {

        let arguments = call.arguments as? [String: Any]
        let callMethod = BKTFlutterCallMethod(rawValue: call.method) ?? .unknown
        switch callMethod {
        case .initialize:
            initialize(arguments, result)

        case .stringVariation:
            stringVariation(arguments, result)

        case .intVariation:
            intVariation(arguments, result)

        case .doubleVariation:
            doubleVariation(arguments, result)

        case .boolVariation:
            boolVariation(arguments, result)

        case .track:
            track(arguments, result)

        case .jsonVariation:
            jsonVariation(arguments, result)

        case .currentUser:
            currentUser(arguments, result)

        case .updateUserAttributes:
            updateUserAttributes(arguments, result)

        case .fetchEvaluations:
            fetchEvaluations(arguments, result)

        case .flush:
            flush(arguments, result)

        case .evaluationDetails:
            evaluationDetails(arguments, result)

        case .addProxyEvaluationUpdateListener:
            // note: will handle in Flutter only. We don't implement native code for these methods
            // see: BucketeerPluginEvaluationUpdateListener.swift
            addProxyEvaluationUpdateListener(result)

        case .destroy:
            do {
                if let listenToken = proxyEvaluationListenToken {
                    try BKTClient.shared.removeEvaluationUpdateListener(key: listenToken)
                    proxyEvaluationListenToken = nil
                }
                try BKTClient.destroy()
                success(result: result)
            } catch {
                fail(result: result, message: error.localizedDescription)
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func success(result: @escaping FlutterResult, response: Any? = nil) {
        let dic = [
            "status": true,
            "response": response
        ] as [String: Any?]
        result(dic)
    }

    func fail(result: @escaping FlutterResult, message: String = "") {
        let dic = [
            "status": false,
            "errorMessage": message
        ] as [String: Any]
        result(dic)
    }
}

File: android/src/main/kotlin/io/bucketeer/sdk/flutter/BucketeerFlutterClientSdkPlugin.kt
package io.bucketeer.sdk.flutter

import android.content.Context
import io.bucketeer.sdk.android.BKTClient
import io.bucketeer.sdk.android.BKTConfig
import io.bucketeer.sdk.android.BKTException
import io.bucketeer.sdk.android.BKTUser
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * BucketeerPlugin
 */
class BucketeerFlutterClientSdkPlugin : MethodCallHandler, FlutterPlugin {
  private var applicationContext: Context? = null
  private var methodChannel: MethodChannel? = null
  override fun onAttachedToEngine(binding: FlutterPluginBinding) {
    onAttachedToEngine(binding.applicationContext, binding.binaryMessenger)
  }

  private fun onAttachedToEngine(applicationContext: Context, messenger: BinaryMessenger) {
    this.applicationContext = applicationContext
    methodChannel = MethodChannel(messenger, "jp.bucketeer.plugin/flutter")
    methodChannel!!.setMethodCallHandler(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
    applicationContext = null
    methodChannel!!.setMethodCallHandler(null)
    methodChannel = null
  }

  private fun initialize(call: MethodCall, result: MethodChannel.Result) {
    val debugging = (call.argument("debugging") as? Boolean) ?: false
    val userId = call.argument("userId") as? String
    val apiKey = call.argument("apiKey") as? String
    val apiEndpoint = call.argument("apiEndpoint") as? String
    val featureTag = call.argument("featureTag") as? String
    val eventsFlushInterval =
      call.argument("eventsFlushInterval") as? Long
    val eventsMaxQueueSize =
      call.argument("eventsMaxQueueSize") as? Int
    val pollingInterval =
      call.argument("pollingInterval") as? Long
    val backgroundPollingInterval =
      call.argument("backgroundPollingInterval") as? Long
    val appVersion = call.argument("appVersion") as? String
    if (apiKey.isNullOrEmpty()) {
      return fail(result, "Missing apiKey")
    }
    if (apiEndpoint.isNullOrEmpty()) {
      return fail(result, "Missing apiEndpoint")
    }
    if (featureTag.isNullOrEmpty()) {
      return fail(result, "Missing featureTag")
    }
    if (userId.isNullOrEmpty()) {
      return fail(result, "Missing userId")
    }
    if (appVersion.isNullOrEmpty()) {
      return fail(result, "Missing appVersion")
    }
    try {
      val config: BKTConfig = BKTConfig.builder()
        .apiKey(apiKey)
        .apiEndpoint(apiEndpoint)
        .featureTag(featureTag).let {
          if (eventsFlushInterval != null && eventsFlushInterval > 0) {
            return@let it.eventsFlushInterval(eventsFlushInterval)
          }
          return@let it
        }.let {
          if (eventsMaxQueueSize != null && eventsMaxQueueSize > 0) {
            return@let it.eventsMaxQueueSize(eventsMaxQueueSize)
          }
          return@let it
        }.let {
          if (pollingInterval != null && pollingInterval > 0) {
            return@let it.pollingInterval(pollingInterval)
          }
          return@let it
        }.let {
          if (backgroundPollingInterval != null && backgroundPollingInterval > 0) {
            return@let it.pollingInterval(backgroundPollingInterval)
          }
          return@let it
        }
        .appVersion(appVersion)
        .build()
      val user: BKTUser = BKTUser.builder().id(userId).build()
      BKTClient.initialize(applicationContext!!, config, user, 5000)
      success(result, true)
    } catch (ex: Exception) {
      fail(result, ex.message)
    }
  }

  private fun currentUser(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val user = BKTClient.getInstance().currentUser()
    val map: MutableMap<String, Any> = HashMap()
    map["id"] = user.id
    map["data"] = user.attributes
    success(result, map)
  }

  private fun evaluationDetails(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val args = call.arguments<Map<String, Any>>()!!
    val featureId = args["featureId"] as? String
      ?: return fail(result, "Missing featureId")
    val evaluation = BKTClient.getInstance().evaluationDetails(featureId)
    if (evaluation == null) {
      fail(result, "Failed to fetch the evaluation.")
    } else {
      val map: MutableMap<String, Any> = HashMap()
      map["id"] = evaluation.id
      map["featureId"] = evaluation.featureId
      map["featureVersion"] = evaluation.featureVersion
      map["userId"] = evaluation.userId
      map["variationId"] = evaluation.variationId
      map["variationValue"] = evaluation.variationValue
      map["reason"] = evaluation.reason.name
      success(result, map)
    }
  }

  private fun stringVariation(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val args = call.arguments<Map<String, Any>>()!!
    val featureId = args["featureId"] as? String
      ?: return fail(result, "Missing featureId")
    val defaultValue = args["defaultValue"] as? String
      ?: return fail(result, "Missing defaultValue")
    val response = BKTClient.getInstance().stringVariation(featureId, defaultValue)
    success(result, response)
  }

  private fun intVariation(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val args = call.arguments<Map<String, Any>>()!!
    val featureId = args["featureId"] as? String
      ?: return fail(result, "Missing featureId")
    val defaultValue = args["defaultValue"] as? Int
      ?: return fail(result, "Missing defaultValue")
    val response = BKTClient.getInstance().intVariation(featureId, defaultValue)
    success(result, response)
  }

  private fun doubleVariation(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val args = call.arguments<Map<String, Any>>()!!
    val featureId = args["featureId"] as? String
      ?: return fail(result, "Missing featureId")
    val defaultValue = args["defaultValue"] as? Double
      ?: return fail(result, "Missing defaultValue")
    val response = BKTClient.getInstance().doubleVariation(featureId, defaultValue)
    success(result, response)
  }

  private fun boolVariation(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val args = call.arguments<Map<String, Any>>()!!
    val featureId = args["featureId"] as? String
      ?: return fail(result, "Missing featureId")
    val defaultValue = args["defaultValue"] as? Boolean
      ?: return fail(result, "Missing defaultValue")
    val response = BKTClient.getInstance().booleanVariation(featureId, defaultValue)
    success(result, response)
  }

  private fun track(call: MethodCall, result: MethodChannel.Result) {
    assertInitialize()
    val args = call.arguments<Map<String, Any>>()!!
    val goalId = args["goalId"] as? String
      ?: return fail(result, "Missing goalId")
    val value = args["value"] as? Double
      ?: return fail(result, "Missing goal value")
    BKTClient.getInstance().track(goalId, value)
    success(result, true)
  }

  private fun jsonVariation(call: MethodCall, result: MethodChannel.Result) {
    try {
      val args = call.arguments<Map<String, Any>>()!!
      val featureId = args["featureId"] as? String
        ?: return fail(result, "Missing featureId")
      val defaultValue = args["defaultValue"] as? Map<String, String> ?: mapOf()
      val response =
        BKTClient.getInstance().jsonVariation(featureId, JSONObject(defaultValue))
      val rawJson = response.toMap()
      success(result, rawJson)
    } catch (ex: Exception) {
      fail(result, message = ex.message ?: "get JsonVariation fail")
    }
  }

  private fun updateUserAttributes(call: MethodCall, result: MethodChannel.Result) {
    val args = call.arguments<Map<String, String>>()!!
    BKTClient.getInstance().updateUserAttributes(args)
    success(result, true)
  }

  private fun fetchEvaluations(call: MethodCall, result: MethodChannel.Result) {
    val args = call.arguments<Map<String, Any>>()!!
    val timeoutMillis = args["timeoutMillis"] as? Long
    MainScope().launch {
      val err = withContext(Dispatchers.IO) {
        return@withContext BKTClient.getInstance().fetchEvaluations(timeoutMillis).get()
      }
      if (err != null) {
        fail(result, err.message)
      } else {
        success(result, true)
      }
    }
  }

  private fun flush(call: MethodCall, result: MethodChannel.Result) {
    MainScope().launch {
      val err = withContext(Dispatchers.IO) {
        return@withContext BKTClient.getInstance().flush().get()
      }
      if (err != null) {
        success(result, err.message)
      } else {
        success(result, true)
      }
    }
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    try {
      when (CallMethods.values().firstOrNull { call.method.lowercase() == it.name.lowercase() }
        ?: CallMethods.Unknown) {
        CallMethods.Initialize -> {
          initialize(call, result)
        }

        CallMethods.CurrentUser -> {
          currentUser(call, result)
        }

        CallMethods.EvaluationDetails -> {
          evaluationDetails(call, result)
        }

        CallMethods.StringVariation -> {
          stringVariation(call, result)
        }

        CallMethods.IntVariation -> {
          intVariation(call, result)
        }

        CallMethods.DoubleVariation -> {
          doubleVariation(call, result)
        }

        CallMethods.BoolVariation -> {
          boolVariation(call, result)
        }

        CallMethods.Track -> {
          track(call, result)
        }

        CallMethods.JsonVariation -> {
         jsonVariation(call, result)
        }

        CallMethods.UpdateUserAttributes -> {
          updateUserAttributes(call, result)
        }

        CallMethods.FetchEvaluations -> {
          fetchEvaluations(call, result)
        }

        CallMethods.Flush -> {
          flush(call, result)
        }

        CallMethods.AddEvaluationUpdateListener -> {
          result.notImplemented()
        }

        CallMethods.RemoveEvaluationUpdateListener -> {
          result.notImplemented()
        }

        CallMethods.ClearEvaluationUpdateListeners -> {
          result.notImplemented()
        }

        CallMethods.Destroy -> {
          BKTClient.destroy()
          success(result, true)
        }

        CallMethods.Unknown -> {
          result.notImplemented()
        }

      }
    } catch (e: BKTException) {
      fail(result, e.message)
    } catch (e: RuntimeException) {
      fail(result, e.message)
    }
  }

  @Throws(IllegalStateException::class)
  private fun assertInitialize() {
    BKTClient.getInstance()
  }

  private fun success(result: MethodChannel.Result?, response: Any? = null) {
    val map: MutableMap<String, Any?> = HashMap()
    map["status"] = true
    map["response"] = response
    result?.success(map)
  }

  private fun fail(result: MethodChannel.Result?, message: String?) {
    val map: MutableMap<String, Any?> = HashMap()
    map["status"] = false
    map["errorMessage"] = message
    result?.success(map)
  }
}

internal enum class CallMethods {
  Initialize,
  StringVariation,
  IntVariation,
  DoubleVariation,
  BoolVariation,
  JsonVariation,
  Track,
  CurrentUser,
  UpdateUserAttributes,
  FetchEvaluations,
  Flush,
  EvaluationDetails,
  AddEvaluationUpdateListener,
  RemoveEvaluationUpdateListener,
  ClearEvaluationUpdateListeners,
  Destroy,
  Unknown
}

@Throws(JSONException::class)
fun JSONObject.toMap(): Map<String, Any> {
  val map = mutableMapOf<String, Any>()
  val keysItr: Iterator<String> = this.keys()
  while (keysItr.hasNext()) {
    val key = keysItr.next()
    var value: Any = this.get(key)
    when (value) {
      is JSONArray -> value = value.toList()
      is JSONObject -> value = value.toMap()
    }
    map[key] = value
  }
  return map
}

@Throws(JSONException::class)
fun JSONArray.toList(): List<Any> {
  val list = mutableListOf<Any>()
  for (i in 0 until this.length()) {
    var value: Any = this[i]
    when (value) {
      is JSONArray -> value = value.toList()
      is JSONObject -> value = value.toMap()
    }
    list.add(value)
  }
  return list
}